### PR TITLE
Run a cron task with a frequency of seconds taking the sql query and updating the timestamp

### PR DESCRIPTION
**Pull request by anandjakhaniya**
**Update config:**
```
cron_job:
  update_parent_timestamp:
    frequency: 60 #frequency to update parent timestamp in second, default 60
    record_update_limit: 100
```

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/ModelDBConstants.java [hunk 1]
```diff
@@ -198,7 +198,7 @@
   // Telemetry Constants
   String TELEMETRY = "telemetry";
   String OPT_IN = "opt_in";
-  String TELEMENTRY_FREQUENCY = "frequency"; // frequency to share data in hours
+  String FREQUENCY = "frequency"; // frequency to share data in hours
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/ModelDBConstants.java]: Let's keep this variable and use another one. They have very different meanings in the two contexts right now, as the comment itself says.

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  public ParentTimestampUpdateCron() {}
+
+  /** The action to be performed by this timer task. */
+  @Override
+  public void run() {
+    LOGGER.info("ParentTimestampUpdateCron wakeup");
+
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      session.beginTransaction();
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java:25]: Don't create long transactions. Also, none of these SQL commands even require a transaction.

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      // Update project timestamp
+      updateProjectByExperimentTimestamp(session);
+
+      // Update experiment timestamp
+      updateExperimentByExperimentRunTimestamp(session);
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java]: Reverse the order. First update the lower elements of the hierarchy, then the top ones.

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+      // Update experiment timestamp
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java]: Remove or update comment.

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      updateDatasetByDatasetVersionTimestamp(session);
+
+      // Update experiment timestamp
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java]: Remove or update comment.

**Review by conradoverta — COMMENTED** [review]
Queries look ok, but please check that we have indices on all the affected operations (grouping, joining and filtering). I don't think we do.

File: backend/config.default.yaml [hunk 6]
```diff
@@ -65,3 +65,7 @@ migration:
 
 feature-flag:
   disabled-authz: true
+
+cron_job:
+  update_parent_timestamp:
+    frequency: 1 #frequency to update parent timestamp in minute, default 1
```
Comment (conradoverta) [inline backend/config.default.yaml]: Can we make the config be seconds by default? It will give us more flexibility.

**Review by conradoverta — COMMENTED** [review]
Actually, can we add a LIMIT in the update requests? I want to avoid locking the database for too long. I'd prefer to have the option to update a small batch more frequently than a large batch less frequently.

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java:25]: These SQL commands are the update operation so for the hibernate we can not execute update operation without the transaction. but ye definitely I can reduce transaction scope and create a new one for each.

Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java:25]: I'm not sure that is correct. I believe hibernate works without transactions. In any case, as long as we put the `LIMIT`, doing a transaction per set of elements seems like a fair trade.

File: backend/src/main/java/ai/verta/modeldb/ModelDBConstants.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/ModelDBConstants.java]: constant reverted and created a new one for frequency

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java:25]: Added `LIMIT` in the query.

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java (hunk 3 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java]: Fixed

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java]: Fixed

File: backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java (hunk 5 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/cron_jobs/ParentTimestampUpdateCron.java]: Fixed

File: backend/config.default.yaml (hunk 6 above)
Comment (anandjakhaniya) [reply backend/config.default.yaml]: Change config with the second and LIMIT where default value of the same is 60s & 100 respectively.

**Comment by anandjakhaniya** [comment]
Fixed all review comments with adding indexes, LIMIT in the query, and other review comment fixes.

**Review by conradoverta — APPROVED** [review]
LGTM.

**Pull request merged**